source/audio/supplier/Mp3Supplier.cpp

```diff
@@ -0,0 +1,245 @@
+	mad_frame_init(&frame);
+	mad_synth_init(&synth);
+
+	// Loop until we are done.
+	while(true)
+	{
+		// If the "next" buffer has filled up, wait until it is retrieved.
+		// Generally try to queue up two chunks worth of samples in it, just
+		// in case NextChunk() gets called twice in rapid succession.
+		unique_lock<mutex> lock(decodeMutex);
+		while(!done && buffer.size() >= BUFFER_CHUNK_SIZE * OUTPUT_CHUNK)
+			condition.wait(lock);
+		// Check if we're done.
+		if(done)
+			break;
+
+		// The lock can be freed until we start filling the output buffer.
+		lock.unlock();
+
+		// See if any input data is left undecoded in the stream. Typically
+		// this is because the last block of input contained a fraction of a
+		// full MP3 frame.
+		size_t remainder = 0;
+		if(stream.next_frame && stream.next_frame < stream.bufend)
+			remainder = stream.bufend - stream.next_frame;
+		if(remainder)
+			memcpy(input.data(), stream.next_frame, remainder);
+
+		// Now, read a chunk of data from the file.
+		data->read(reinterpret_cast<char *>(input.data() + remainder), INPUT_CHUNK - remainder);
+		// If you get the end of the file, loop around to the beginning.
+		if(data->eof() && looping)
+			data->seekg(0, iostream::beg);
+		else if(data->eof())
+		{
+			done = true;
+			// Make sure there are full chunks of data available
+			buffer.resize(OUTPUT_CHUNK * ceil(static_cast<double>(buffer.size()) / static_cast<double>(OUTPUT_CHUNK)));
+		}
+
+		// If there is nothing to decode, return to the top of this loop.
+		if(!(remainder || !data->eof()))
+			continue;
+
+		// Hand the input to the stream decoder.
+		mad_stream_buffer(&stream, &input.front(), INPUT_CHUNK);
+
+		// Loop through the decoded result for that input block.
+		while(true)
+		{
+			// Decode the next frame, and check if there is an error.
+			if(mad_frame_decode(&frame, &stream))
+			{
+				// For recoverable errors, keep going.
+				if(MAD_RECOVERABLE(stream.error))
+					continue;
+				else
+					break;
+			}
+			// Convert the decoded audio into a PCM signal.
+			mad_synth_frame(&synth, &frame);
+
+			// If the source is mono, read both output channels from the left input.
+			// Otherwise, read two separate input channels.
+			mad_fixed_t *channels[2] = {
+					synth.pcm.samples[0],
+					synth.pcm.samples[synth.pcm.channels > 1]
+			};
+
+			// For this part, we need access to the output buffer.
+			lock.lock();
+			if(done)
+				break;
+
+			// We'll alternate what channel we read from each time through the loop.
+			bool channel = false;
+			for(unsigned i = 0; i < 2 * synth.pcm.length; ++i)
+			{
+				// Read the next sample from the next channel.
+				mad_fixed_t sample = *channels[channel]++;
+				channel = !channel;
+
+				// Clip and scale the sample to 16 bits.
+				sample += (1L << (MAD_F_FRACBITS - 16));
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+				sample = max(-MAD_F_ONE, min(MAD_F_ONE - 1, sample));
+#pragma GCC diagnostic pop
```

Comment:
I know this was in the current code, but since you're rewriting the whole audio system anyway...

Where is the old cast?